independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/PathTreeClassPathElement.java

```diff
@@ -0,0 +1,257 @@
+
+public class PathTreeClassPathElement extends AbstractClassPathElement {
+
+    static class ZipFileMayHaveChangedException extends RuntimeException {
+        public ZipFileMayHaveChangedException(Throwable cause) {
+            super(cause);
+        }
+    }
+
+    static {
+        //force this class to be loaded
+        //if quarkus is recompiled it needs to have already
+        //been loaded
+        //this is just a convenience for quarkus devs that means exit
+        //should work properly if you recompile while quarkus is running
+        new ZipFileMayHaveChangedException(null);
+    }
+
+    private static final Logger log = Logger.getLogger(PathTreeClassPathElement.class);
+
+    private final Lock readLock;
+    private final Lock writeLock;
+    private final OpenPathTree pathTree;
+    private final boolean runtime;
+    private final ArtifactKey dependencyKey;
+    private volatile boolean closed;
+    private volatile Set<String> resources;
+
+    public PathTreeClassPathElement(PathTree pathTree, boolean runtime) {
+        this(pathTree, runtime, null);
+    }
+
+    public PathTreeClassPathElement(PathTree pathTree, boolean runtime, ArtifactKey dependencyKey) {
+        this.pathTree = Objects.requireNonNull(pathTree, "Path tree is null").openTree();
+        final ReadWriteLock readWriteLock = new ReentrantReadWriteLock();
+        this.readLock = readWriteLock.readLock();
+        this.writeLock = readWriteLock.writeLock();
+        this.runtime = runtime;
+        this.dependencyKey = dependencyKey;
+    }
+
+    @Override
+    public boolean isRuntime() {
+        return runtime;
+    }
+
+    @Override
+    public ArtifactKey getDependencyKey() {
+        return dependencyKey;
+    }
+
+    @Override
+    public Path getRoot() {
+        return pathTree.getOriginalTree().getRoots().iterator().next();
+    }
+
+    @Override
+    public ClassPathResource getResource(String name) {
+        if (!getProvidedResources().contains(name)) {
+            return null;
+        }
+        return withOpenTree(tree -> tree.processPath(name, new ClassPathResourceProducer()));
+    }
+
+    @Override
+    public <T> T withOpenTree(Function<OpenPathTree, T> func) {
+        readLock.lock();
+        try {
+            if (closed) {
+                //we still need this to work if it is closed, so shutdown hooks work
+                //once it is closed it simply does not hold on to any resources
+                try (OpenPathTree openTree = pathTree.getOriginalTree().openTree()) {
+                    return func.apply(openTree);
+                } catch (IOException e) {
+                    throw new UncheckedIOException(e);
+                }
+            } else {
+                return func.apply(pathTree);
+            }
+        } finally {
+            readLock.unlock();
+        }
+    }
+
+    @Override
+    public Set<String> getProvidedResources() {
+        if (resources == null) {
+            synchronized (this) {
```

Comment:
I'm probably missing something important or obvious, but shouldn't this use the locks instead?

Reply:
Good point.